src/charm.py

```diff
@@ -1,104 +1,400 @@
+
+        container = self.unit.get_container(self._pgbouncer_service)
+        layer = self._pgbouncer_layer()
+
+        if not container.can_connect():
+            self.unit.status = WaitingStatus("waiting for pebble in pgbouncer workload container.")
+            return
+
+        services = container.get_plan().to_dict().get("services", {})
+        if services != layer["services"]:
+            container.add_layer(self._pgbouncer_service, layer, combine=True)
+            logging.info("Added layer 'pgbouncer' to pebble plan")
+            container.restart(self._pgbouncer_service)
+            logging.info("restarted pgbouncer service")
+        self.unit.status = ActiveStatus()
+
+    def _pgbouncer_layer(self) -> Layer:
+        """Returns a default pebble config layer for the pgbouncer container."""
+        return {
+            "summary": "pgbouncer layer",
+            "description": "pebble config layer for pgbouncer",
             "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
+                self._pgbouncer_service: {
+                    "summary": "pgbouncer service",
+                    "user": self._pgbouncer_user,
+                    "command": f"pgbouncer {INI_PATH}",
                     "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
+                    "override": "replace",
+                    "environment": {
+                        "PGB_DATABASES": self.config["pgb_databases"],
+                        "PGB_LISTEN_PORT": self.config["pgb_listen_port"],
+                        "PGB_LISTEN_ADDRESS": self.config["pgb_listen_address"],
+                        "PGB_ADMIN_USERS": self.config["pgb_admin_users"],
+                    },
                 }
             },
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
+
+    def _on_pgbouncer_pebble_ready(self, event) -> None:
+        """Define and start pgbouncer workload."""
+        container = event.workload
+        pebble_layer = self._pgbouncer_layer()
+        container.add_layer(self._pgbouncer_service, pebble_layer, combine=True)
         container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
         self.unit.status = ActiveStatus()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+    # ====================
+    #  Charm Action Hooks
+    # ====================
+
+    def _on_reload_pgbouncer_action(self, event) -> None:
+        """An action to explicitly reload pgbouncer.
+
+        TODO implement function stub
```

Comment:
Yes, all the TODO comments are for future PRs. I'll extract the salient information into JIRA tickets and the spec document for most of them, but I believe this one in particular can remain since it's short and accurately describes this function as a stub. It's also the next piece of work I intend to do in this repo.

Reply:
I'll address your comments about this function in this comment. When I say "config", I mean the pgbouncer configuration files (pgbouncer.ini, and userlist.txt), not the charm config. 

This function is an unnecessary stub in itself, but it's a necessary part of the config management, since changes in config are not applied until the pgbouncer _application_ is reloaded. Therefore, it's essential in the unit tests for config management functions that we assert this function has been called, even though it currently doesn't do anything. The unit test related to this function has been removed.

Reply:
I trust you here! This sounds good to me, thanks for the explanation! :D